src/Microsoft.Diagnostics.Runtime/ClrHeap.cs

```diff
@@ -976,6 +990,43 @@ public override IEnumerable<ulong> EnumerateObjectAddresses()
             }
         }
 
+
+        public override IEnumerable<Tuple<ulong, ulong, ClrType, ulong>> EnumerateObjectDetails()
+        {
+            ClrHeap heap = _segments[0].Heap;
+
+            // optimization for large dumps/apps with 30.000+ types:
+            //    keep track of ClrType given a method table address
+            Dictionary<ulong, ClrType> typeCache = new Dictionary<ulong, ClrType>(32768);
+
+            for (int iSegment = 0; iSegment < _segments.Length; ++iSegment)
+            {
+                ClrSegment segment = _segments[iSegment];
+
+                // unlike the default implementation, we are trying to avoid
+                // duplicate calls to GetObjectType and GetSize that are used by segment.NextObject()
+                // --> 25% faster on large memory dumps
```

Comment:
That's a pretty surprising finding.  It's been a loooong time since I've profiled this (so are probably right), but the original intent of some of the implementation of ClrHeap.GetObjectType is to cache the previous call, and if the MethodTable is known we do a dictionary lookup anyway.  I'm surprised that this provides value outside of increasing the dictionary size for methodtables.

On the one hand, this makes me think that the solution here might not be to add another public method, but instead for me to figure out the underlying source of why this would be a 25% perf improvement and fix the underlying perf issue.

On the other hand, I'm actually not opposed to a more detailed EnumerateHeap method with these tradeoffs, assuming there's not something I can tweak in the current implementation to fix the perf problem you report here.

Reply:
How do you stop things like this from regressing without performance tests? We could add some microbenchmarks with BDN (@adamsitnik )

Reply:
Agreed.  ClrMD is in dire need of more tests, perf and otherwise.  :)

Reply:
@davidfowl: I've added a test just for this purpose
@leculver: caching in ClrMD is... complicated and I already found an issue related to caching (https://github.com/Microsoft/clrmd/issues/108). I did not feed confident enough to touch it: this is why I preferred to implement the dedicated new method